tests/test_registries.py

```diff
@@ -0,0 +1,166 @@
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+
+import unittest
+
+import fixtures
+import testtools
+
+from . import base  # noqa: H304
+
+import windlass.registries
+
+
+class TestRegistries(testtools.TestCase):
+
+    def test_from_url_aws_ecr(self):
+        registry_url = "000000000000.dkr.ecr.us-west-1.amazonaws.com"
+        registry = windlass.registries.from_url(registry_url)
+
+        self.assertIsInstance(registry, windlass.registries.EcrDockerRegistry)
+
+    def test_from_url_jfrog(self):
+        registry_url = "example-org.jfrog.io"
+        registry = windlass.registries.from_url(registry_url)
+
+        self.assertIsInstance(
+            registry, windlass.registries.JfrogDockerRegistry
+        )
+
+    def test_from_url_hub(self):
+        registry_url = "registry.hub.docker.com"
+        registry = windlass.registries.from_url(registry_url)
+
+        self.assertIsInstance(registry, windlass.registries.DockerRegistry)
+
+    def test_from_url_other(self):
+        registry_url = "registry.quay.io"
+        registry = windlass.registries.from_url(registry_url)
+
+        self.assertIsInstance(registry, windlass.registries.DockerRegistry)
+
+
+class TestDockerRegistry(testtools.TestCase):
+
+    reg_url = "registry.hub.docker.com"
+
+    @base.addfixture(fixtures.EnvironmentVariable, 'DOCKER_USER', 'my_user')
+    @base.addfixture(fixtures.EnvironmentVariable, 'DOCKER_PASSWORD',
+                     'my_password')
+    def test_credentials_from_env(self):
+        registry = windlass.registries.DockerRegistry(self.reg_url)
+
+        self.assertEqual('my_user', registry.username)
+        self.assertEqual('my_password', registry.password)
+
+    @base.addfixture(fixtures.EnvironmentVariable, 'DOCKER_USER', 'my_user')
+    @base.addfixture(fixtures.EnvironmentVariable, 'DOCKER_PASSWORD',
+                     'my_password')
+    def test_credentials_ignore_env_when_set(self):
+        registry = windlass.registries.DockerRegistry(
+            self.reg_url,
+            "explicit_user",
+            "explicit_password",
+        )
+
+        self.assertEqual('explicit_user', registry.username)
+        self.assertEqual('explicit_password', registry.password)
+
+
+@unittest.mock.patch('windlass.registries.remotes.ECRConnector')
+class TestEcrRegistry(testtools.TestCase):
+
+    reg_url = "000000000000.dkr.ecr.us-west-1.amazonaws.com"
+
+    @base.addfixture(fixtures.EnvironmentVariable, 'AWS_ACCESS_KEY_ID',
+                     'AKIAIOSFODNN7EXAMPLE')
+    @base.addfixture(fixtures.EnvironmentVariable, 'AWS_SECRET_ACCESS_KEY',
+                     'wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY')
+    @base.addfixture(fixtures.EnvironmentVariable, 'AWS_DEFAULT_REGION',
```

Comment:
This one could be inside a test. while ones defined in 88-90 and 105-107 could go to setUp.

Reply:
Good point